### PR TITLE
DAOS-10339 pool: Fix pool creation drpc response

### PR DESCRIPTION
**Pull request by knard-intel**
# Description

Fix fields of the pool creation drpc reponse.  Some of these fields where not filled by the engine: the control plane was using value from the request which could be different from what was really done.

### CODE REVIEW

File: src/tests/ftest/pool/create_query.py
Comment (daltonbohning) [inline src/tests/ftest/pool/create_query.py:34]: Minor - Namespace is not required if you're just passing the default
```suggestion
        self.add_pool()
```

File: src/tests/ftest/pool/create_query.py
Comment (daltonbohning) [inline src/tests/ftest/pool/create_query.py:31]: Optional/Recommendation - Personally, I like to use a unique tag that includes the directory/feature and filename/test case.
```suggestion
        :avocado: tags=pool_create_query
```
That way it's intuitive that `pool_create_query` is in the `pool` directory and runs the `create_query` test. But this is not a requirement.

File: src/tests/ftest/pool/create_query.py
Comment (daltonbohning) [inline src/tests/ftest/pool/create_query.py:14]: Minor - According to the config, it's just 4 servers. And not really a "client" since it's all dmg.
```suggestion
    All of the tests verify pool create performance with 4 servers.
```
Also, is this checking "performance" in some way? (Maybe just copy-pasted)

File: src/tests/ftest/pool/create_query.yaml
Comment (daltonbohning) [inline src/tests/ftest/pool/create_query.yaml:22]: Minor - It's preferred to use human-readable format when possible
```suggestion
  scm_size: 10G
  nvme_size: 100G
```

File: src/tests/ftest/pool/create_query.py
Comment (daltonbohning) [inline src/tests/ftest/pool/create_query.py:39]:
```suggestion
        self.assertGreaterEqual(self.pool.scm_per_rank, self.pool.scm_size.value,
                "SCM size of the pool created should be greater than or equal to the given size")
```

**Review by daltonbohning — COMMENTED** [review]
Mostly minor comments for ftest changes

File: src/tests/ftest/pool/create_query.py
Comment (kjacque) [inline src/tests/ftest/pool/create_query.py:15]: Is there a test tag for NVMe drives and unsimulated pmem? I don't know if "hw" always implies real storage.

**Review by kjacque — APPROVED** [review]

File: src/tests/ftest/pool/create_query.py
Comment (daltonbohning) [reply src/tests/ftest/pool/create_query.py:15]: There is a tag `nvme`, but not a similar `pmem` as far as I can tell. And `hw` just means it runs on a hardware cluster instead of VM. Hardware clusters always have real hardware, but the test could theoretically not use it.

Comment (kjacque) [reply src/tests/ftest/pool/create_query.py:15]: Seems like this test should at least tag for nvme then. So by real hardware you mean all the CI nodes have physical NVDIMMs? As stated here, that seems to be a requirement for this test.

**Review by tanabarr — DISMISSED** [review]

File: src/tests/ftest/pool/create_query.py
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:34]: - [ ] Remove useless param

File: src/tests/ftest/pool/create_query.py
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:31]: - [ ] Fix tag name

Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:31]: Initially it was in the pool test file, and I used its naming convention. However, I agree that your proposition is more intuitive (at least for a DAOS newbie such as me)

File: src/tests/ftest/pool/create_query.py
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:14]: Copy paste is the direct way to hell :'(

- [ ] Fix hardware description

File: src/tests/ftest/pool/create_query.yaml
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.yaml:22]: - [ ]  Fix pool size definiton

File: src/tests/ftest/pool/create_query.py
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:39]: - [ ] Fix spelling

File: src/tests/ftest/pool/create_query.py
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.py:15]: I confirm, that having PMEM and and NVMe is needed for this test as we want to check that the size allocated on these devices are consistent with the input parameters.

File: src/tests/ftest/pool/create_query.yaml
Comment (knard-intel) [reply src/tests/ftest/pool/create_query.yaml]: At  the end the human readable is not usable as it is also used to check the size of the created pool: The CI failed asking that string could not compared with int.

File: src/tests/ftest/pool/create_query.py
Comment (kjacque) [reply src/tests/ftest/pool/create_query.py:15]: Superficially skimming through it, it looks like NVMe nodes generally have physical PMEM. So the nvme tag may be sufficient for now.

File: src/mgmt/srv_drpc.c
Comment (mjmac) [inline src/mgmt/srv_drpc.c:379]: Is there a potential race here between the pool service having been created and being ready to handle queries? Just concerned about pool create operations appearing to fail because this query fails, leading to a pool cleanup and confusing behavior for the user.

File: src/mgmt/srv_drpc.c
Comment (mjmac) [inline src/mgmt/srv_drpc.c:399]: Seems like it would be better to do this in a loop, e.g. `for (i = 0; i < DAOS_MEDIA_MAX; i++)` in order to be future-proof for the addition of new tier types.

File: src/mgmt/srv_drpc.c
Comment (knard-intel) [reply src/mgmt/srv_drpc.c:379]: I agree, that there is a potential race condition as for example for the conversion of the ranklist convertion or the message packing. FMHO,, I do not thing that it will be confusing for the end user as it will just see that the pool creation as failed. Moreover, such situation should be very uncommon and should probably indicate that the pool created is unusable by the end user. To mitigate this race condition, I could also perform some retries  with random latencies between trials. By this way, we could handle transient networking issue.

File: src/mgmt/srv_drpc.c
Comment (knard-intel) [reply src/mgmt/srv_drpc.c:399]: I hesitated between a loop or doing it manually, and I finally preferred the second solution as it was simple: KISS principle ;-) However, I agree that doing it with the loop as you suggested is more future-proof.

- [ ] do tier_bytes assignment within a loop

File: src/mgmt/srv_drpc.c
Comment (mjmac) [reply src/mgmt/srv_drpc.c:379]: Well, I was under the impression that the implementation of `ds_mgmt_pool_query()` might handle transient communication issues with retries, but I'm not sure. I think it's worth confirming either way though. Mostly what I want to avoid is a situation where we do the pool query here immediately after creating it, and having a failure that causes the pool to be automatically cleaned up. 

@kccain, @liw: Any advice here?

File: src/mgmt/srv_drpc.c
Comment (mjmac) [reply src/mgmt/srv_drpc.c:399]: I don't see a downside to doing it in a loop. We already have a lot of technical debt due to old assumptions about only having two storage tiers, but we should be writing new code with the understanding that we are planning to support N storage tiers.

File: src/mgmt/srv_drpc.c
Comment (kccain) [reply src/mgmt/srv_drpc.c:379]: A pool query to a pool service that is not ready yet should (I think) result in a cart RPC reply with status -DER_NOTLEADER and a hint of what rank to try to contact next in the "search for the leader" logic. ds_mgmt_pool_query() (though running on an engine) does have client RPC logic to rechoose another pool service replica rank and retry the RPC when that -DER_NOTLEADER response is seen. And this should continue until a successful reply, or some other error is seen (one that I assume would be unrelated to any pool start up latency).

Some questions come to mind though. In cases where scm size is specifically given by the admin in the pool create request was it ever promised (from an API/interface perspective) that the exact number of bytes allocated would be returned in the pool create response? In other cases where a single size is given, and scm and nvme sizes are calculated based on a ratio I guess it may make more sense to return the calculated (or maybe even actual allocated) amount to be informative to the admin about what was allocated. In the scm_size given in the request, I see in the code why a subsequent list-pools/query may report a different amount, given the single target size calculation that rounds up to a multiple of 4KB.

**Review by kccain — COMMENTED** [review]
responding to the one question, haven't studied the full patch yet

File: src/mgmt/srv_drpc.c
Comment (liw) [inline src/mgmt/srv_drpc.c:447]: [Nit] Since the "pool_create_fill_resp" name does not imply that the function consumes the svc parameter, it would be normal to still free svc in ds_mgmt_drpc_pool_create.

File: src/mgmt/srv_drpc.c
Comment (liw) [reply src/mgmt/srv_drpc.c:379]: Although using a pool query may not be the most economical way (for we could gather the info via the responses of the RPCs that create the pool, I think it should work.

Comment (knard-intel) [reply src/mgmt/srv_drpc.c:379]: From what I remember when I did this stuff, the call to the RPC `ds_mgmt_create_pool()` and its subsequent function call does not return the really allocated size. Thus, I used the pool query to avoid too much intrusive code modification. However, from what I understand, such modifications are worthwhile and thus I will update my PR to not have to do some extra RPC call.

- [ ] Remove RPC poll_query call

File: src/mgmt/srv_drpc.c
Comment (knard-intel) [reply src/mgmt/srv_drpc.c:447]: - [ ] Free svc parameter in `ds_mgmt_drpc_pool_create()` function body.

File: src/mgmt/srv_drpc.c
Comment (liw) [reply src/mgmt/srv_drpc.c:379]: I missed a ")" in my last comment. Added. The current pool-query-based approach is fine with me.

Comment (knard-intel) [reply src/mgmt/srv_drpc.c:379]: After more investigation with @mjmac, we agreed that the pool-query-base was not the best approach (in terms of performance) but was acceptable to fix this issue: it should not introduce reliability issue.

**Review by liw — APPROVED** [review]
Looks good to me. Thanks for fixing the "[Nit]". Might make sense to wait for a more through review from a control plane expert.

**Review by mjmac — APPROVED** [review]

**Pull request merged**
